mParticle-Apple-SDK/MPBackendController.m

```diff
@@ -965,161 +772,55 @@ - (void)handleNetworkPerformanceNotification:(NSNotification *)notification {
-    __weak MPBackendController *weakSelf = self;
-    
-    dispatch_async(dispatch_get_main_queue(), ^{
-        __strong MPBackendController *strongSelf = weakSelf;
-        if (!strongSelf) {
-            return;
-        }
-        
-        if (strongSelf->uploadSource) {
-            dispatch_source_cancel(strongSelf->uploadSource);
-            strongSelf->uploadSource = nil;
-        }
-        
-        strongSelf->uploadSource = [strongSelf createSourceTimer:strongSelf.uploadInterval
-                                                    eventHandler:^{
-                                                        dispatch_async([MParticle messageQueue], ^{
-                                                            __strong MPBackendController *strongSelf = weakSelf;
-                                                            if (!strongSelf) {
-                                                                return;
-                                                            }
-                                                            [strongSelf waitForKitsAndUploadWithCompletionHandler:nil];
-                                                        });
-                                                        
-                                                    } cancelHandler:^{
-                                                        
-                                                    }];
-    });
-}
 
+// Timer blocks fire on message queue
 - (dispatch_source_t)createSourceTimer:(uint64_t)interval eventHandler:(dispatch_block_t)eventHandler cancelHandler:(dispatch_block_t)cancelHandler {
-    dispatch_source_t sourceTimer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, dispatch_get_main_queue());
-    
+    dispatch_source_t sourceTimer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, [MParticle messageQueue]);
+
     if (sourceTimer) {
         dispatch_source_set_timer(sourceTimer, dispatch_walltime(NULL, 0), interval * NSEC_PER_SEC, 0.1 * NSEC_PER_SEC);
         dispatch_source_set_event_handler(sourceTimer, eventHandler);
         dispatch_source_set_cancel_handler(sourceTimer, cancelHandler);
         
-        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(interval * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(interval * NSEC_PER_SEC)), [MParticle messageQueue], ^{
             dispatch_resume(sourceTimer);
         });
     }
     
     return sourceTimer;
 }
 
-- (void)endBackgroundTimer {
-    if (backgroundSource) {
-        dispatch_source_cancel(backgroundSource);
+- (void)beginUploadTimer {
+    @synchronized (self) {
+        if (self.uploadSource) {
+            dispatch_source_cancel(self.uploadSource);
+            self.uploadSource = nil;
+        }
+        
+        self.uploadSource = [self createSourceTimer:self.uploadInterval eventHandler:^{
+            [self waitForKitsAndUploadWithCompletionHandler:nil];
```

Comment:
Just checking - I see we're calling waitForKitsAndUpload (which will make batches) within `beginUploadTimer`, which is called within `applicationDidBecomeActive` - does an app become "inactive" only after the countdown timer goes to zero? If so, then this is fine.

or does the app not become inactive